File: d01/main.py
#!/usr/bin/env python3

from collections import Counter
from sys import argv
from time import process_time

tasks: list[int] = [1, 2]
test_only: bool = True

def solve():
    inputf = "input_testdata.txt" if test_only else "input.txt"

    start = process_time()

    puzzle = open(inputf).read()
    all_ints = [int(n) for n in puzzle.split()]
    a = sorted(all_ints[::2])
    b = sorted(all_ints[1::2])

    load = process_time()

    if 1 in tasks:
        # task 1: sum up the absolute differences of the values
        print(sum([abs(x-y) for x, y in zip(a, b)]))
    
    if 2 in tasks:
        # task 2: sum up the product of each value in list 1 multiplied by number of occurence in list 2
        counter = Counter(b)
        totals = sum([v * counter[v] for v in a])
        print(totals)

    stop = process_time()

    print(f"Load took {load-start:3.6f} secs, calculation {stop-load:3.6f} secs")


def main(argv):
    global test_only, tasks

    if "-h" in argv or "--help" in argv:
        print("usage: python3 main.py [-h] [-s] [-1|2]")
        print("  -h:    this help")
        print("  -s:    execute large scale problem, not just example/test set, which is the default")
        print("  -1|2:  task 1 or 2, respectively (default: both tasks)")
        exit(1)

    tasks = [2] if "-2" in argv else [1] if "-1" in argv else [1,2]

    test_only = False if "-s" in argv else True

    print(f"\n===> Running task(s) {tasks} in {'test' if test_only else 'full'} mode ===================================\n")

    solve()


if __name__ == "__main__":
    main(argv)

    # read, eg, 5,6, and return as list of int tupels
    # input = "5,6\n3,4\n2,3"
    # lst = list(tuple((int(x), int(y))) for x, y in re.findall('([0-9]+),([0-9]+)', input))

    # read alphaanumeric (incl '_'), and return as list of lists
    # input = "key -> value\nk2 -> val_2\nk3 -> v3"
    # lst = list( [ x, y ] for x, y in re.findall('([\w]+) -> ([\w]+)', input))
    # print (lst)

    # ... or as dict
    # dct = dict( [ x, y ] for x, y in re.findall('([\w]+) -> ([\w]+)', input))
    # print(dct)

File: d01/main.js
#!/usr/bin/env node

// Note: I am just learning JavaScript (not from "ground up", I know several languages like C, C++, Java, Python, ...),
// but more like how to work best with lists, dictionaires, etc., and getting "hands on experience"
// You will find a lot of boilerplate code here, and also some of my solutions might suffer from the fact that
// I am still learning. You have been warned... ;-)

import { readFileSync } from 'node:fs';
import { exit } from 'node:process';

function load_data(test_only) {
    const dir = import.meta.dirname
    if (test_only) {
        return readFileSync(dir + '/input_testdata.txt', 'utf8')
    } else {
        return readFileSync(dir + '/input.txt', 'utf8')
    }
}

if (process.argv.includes("-h") || process.argv.includes("--help")) {
    console.log("usage: node main.js [-h] [-s] [-1|2]")
    console.log("  -h:    this help")
    console.log("  -s:    execute large scale problem, not just example/test set, whiich is the default")
    console.log("  -1|2:  task 1 or 2 only, repectively (default: both tasks)")
    exit(1)
}

let tasks = process.argv.includes("-2") ? [2] : process.argv.includes("-1") ? [1] : [1, 2]
let test_only = process.argv.includes("-s") ? false : true

let start = performance.now()

// As part of my "boilerplate" I read data into an array of lines of strings
// and keep track of the dimension (m = number of columns, n = number of rows)

let data = load_data(test_only)
let lines = data.trim().split('\n')
let n = lines.length
let m = lines[0].length
console.log(`Size: ${n}x${m}`)

let load = performance.now()

// Actual problem solving starts here ===========================================================================

// This day, I do not need the lines, but can read the numbers of the two columns directly from the data,
// and then sorting by their values - learning about JavaScript regex as well as map() on day 1 - yeah!
let all_numbers = data.trim().split(/ +|\n/).map(Number)
let a = all_numbers.filter((_elem, idx) => { return idx%2 == 0 }).sort()
let b = all_numbers.filter((_elem, idx) => { return idx%2 == 1 }).sort()

// a helper class I used in part 2 - inspired by Pythons Counter class
class Counter {
    constructor(array) {
        this.dict = {}
        for (const a of array) {
            this.dict[a] = (this.dict[a] ?? 0) + 1
        }
    }

    val(num) {
        return this.dict[num] ?? 0
    }
}

// Part 1 is easy by just iterating over the two columns and adding up their 
// absolute differences

if (tasks.includes(1)) {

    let totals = 0
    for (let i = 0; i < a.length; i++) {
        totals += Math.abs(a[i]-b[i]) 
    }
    console.log(`Part 1: ${totals}`)    
}

// Part 2 leverages the Counter class above to count occurences in the second column
// and then reducing a with it

if (tasks.includes(2)) {
    let counter = new Counter(b)
    let totals = a.reduce((accumulator, current) => accumulator+(current * counter.val(current)), 0)
    console.log(`Part 2: ${totals}`)    
}

let stop = performance.now()

console.log(`Load took ${((load-start)/1000).toFixed(6)} msecs, calculation ${((stop-load)/1000).toFixed(6)} msecs`)
